public/app/plugins/datasource/loki/components/useLokiLabels.ts

```diff
@@ -0,0 +1,75 @@
+import { useState, useEffect } from 'react';
+import LokiLanguageProvider from 'app/plugins/datasource/loki/language_provider';
+import { TypeaheadInput } from 'app/features/explore/QueryField';
+
+/**
+ *
+ * @param languageProvider
+ * @param languageProviderInitialised
+ * @param activeOption rc-cascader provided option used to fetch option's values that hasn't been loaded yet
+ *
+ * @description Fetches missing labels and enables labels refresh
+ */
+export const useLokiLabels = (
+  languageProvider: LokiLanguageProvider,
+  languageProviderInitialised: boolean,
+  activeOption: TypeaheadInput[]
+) => {
+  // State
+  const [logLabelOptions, setLogLabelOptions] = useState([]);
+  const [shouldTryRefreshLabels, setRefreshLabels] = useState(false);
+
+  // Async
+  const fetchOptionValues = async option => {
+    await languageProvider.fetchLabelValues(option);
+    setLogLabelOptions(languageProvider.logLabelOptions);
+  };
+
+  const tryLabelsRefresh = async () => {
+    await languageProvider.refreshLogLabels();
+    setRefreshLabels(false);
+    setLogLabelOptions(languageProvider.logLabelOptions);
+  };
+
+  // Effects
+
+  // This effect performs loading of options that hasn't been loaded yet
+  // It's a subject of activeOption state change only. This is because of specific behavior or rc-cascader
+  // https://github.com/react-component/cascader/blob/master/src/Cascader.jsx#L165
+  useEffect(() => {
+    if (!languageProviderInitialised) {
```

Comment:
Minor, rename all `initialised` => `initialized`.
```suggestion
    if (!languageProviderInitialized) {
```